1. Realizar una reducción de dimensionalidad/PCA para intentar aumentar el porcentaje de acierto. ¿Aumenta? ¿Cómo cambia el acierto según diferentes valores de PCA? Prueba, como mínimo, con valores para "n_components" 2,4,5 y 0.45 y 0.95. ¿Ha habido algún problema con algún valor?

In [2]:
import pandas as pd

In [6]:
### for this exercise I used my pca_df function in Utils. for each n_components I tried, I saved the score of the test set. Then I created a df in which we can compare the scores of the different models with the differen n_components. 

In [5]:

models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]


data = {"pca_2":[73.74, 73.74, 79.89, 36.31, 67.6, 73.74, 74.86, 83.8, 83.8], "pca_4": [81.01, 82.12, 80.45, 67.04, 79.89, 81.01, 74.3, 83.24, 83.24], "pca_5": [80.45, 80.45, 79.33, 69.27, 77.09, 79.89, 81.56, 82.68, 84.36],  "pca_0.45": [65.36, 65.36, 79.33, 36.31, 48.04, 67.6, 48.04, 82.68, 82.68], "pca_0.95": [73.74, 73.74, 79.89, 36.31, 67.6, 73.74, 73.18, 83.8, 83.8]}

df_scores = pd.DataFrame(data, index = models)
df_scores.index.name = 'model'

df_scores.sort_values(by='pca_4', ascending=False)

pca_2  pca_4  pca_5  pca_0.45  pca_0.95
model                                                 
decision_tree  83.80  83.24  82.68     82.68     83.80
random_forest  83.80  83.24  84.36     82.68     83.80
SVC            73.74  82.12  80.45     65.36     73.74
logreg         73.74  81.01  80.45     65.36     73.74
LinearSVC      73.74  81.01  79.89     67.60     73.74
knn            79.89  80.45  79.33     79.33     79.89
perceptron     67.60  79.89  77.09     48.04     67.60
sgdclassifier  74.86  74.30  81.56     48.04     73.18
gaussian       36.31  67.04  69.27     36.31     36.31

In [8]:
for elem in df_scores.columns:
    mean = df_scores[elem].mean()
    print("score mean of", elem, "is: ", mean)

score mean of pca_2 is:  71.94222222222223
score mean of pca_4 is:  79.14444444444445
score mean of pca_5 is:  79.45333333333333
score mean of pca_0.45 is:  63.933333333333344
score mean of pca_0.95 is:  71.75555555555555


The reduction of the dimensions did not increase the accuracy score. The highest score with the original settings was:	86.76 using Random Forest. The highest score we got with pca was 84.36 with Random Forest with n_components 5. 

In general the best scores (mean of scores of all models) were with a pca n_components of 5. The lowest scores were with n_components of 0.45. 

-> to ask: why not highest score with pca_0.95? done correctly?  

2. Realizar una normalización de los datos no categóricos y comprobar si aumenta los scores.

In [ ]:
## used function --> normalize_dataframe(df=test_df, column="Fare") from Utils.cleaning_function to normalize the columns: Fare. This is the only numeric column in the dataframe. 

In [18]:
models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]

scores = [82.12, 84.36, 85.47, 82.68, 68.72, 82.68, 81.01, 78.77, 84.36]

df_scores_2 = pd.DataFrame(scores, index = models)
df_scores_2 = df_scores_2. rename(columns={0:'Score'})
df_scores_2.sort_values(by='Score', ascending=False)

Score
knn            85.47
SVC            84.36
random_forest  84.36
gaussian       82.68
LinearSVC      82.68
logreg         82.12
sgdclassifier  81.01
decision_tree  78.77
perceptron     68.72

The overall general score did not increase. without the change the highest score was: Random Forest -	86.76, now the highest score is knn with 85.47.  But the scores are higher then when using the pca. 

Also the best model without the changes and with pca was: random_forest, and with the normalization the best model is: knn. 

3. Añade, como columnas, la media, mediana, varianza y covarianza de todas las columnas numéricas no categóricas y comprueba si con esas nuevas columnas se aumenta el score de los algoritmos.



In [ ]:
# following steps made:

#train_df["mean_fare"] = train_df.Fare.mean()
#test_df["mean_fare"] = test_df.Fare.mean()

#train_df["median_fare"] = train_df.Fare.median()
#test_df["median_fare"] = test_df.Fare.median()

#train_df["variance_fare"] = train_df.var()['Fare'] 
#test_df["variance_fare"] = test_df.var()['Fare'] 


#--> to ask: .cov gives only nan values? So in this case I did not add the covariance. I tried it like this:
#train_df["covariance_fare"] = train_df.cov()['Fare'] 
#test_df["covariance_fare"] = test_df.cov()['Fare']

In [22]:
models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]

scores = [82.12, 65.36, 75.98, 82.68, 37.99, 65.92, 37.99, 79.89, 84.92]


df_scores_3 = pd.DataFrame(scores, index = models)
df_scores_3 = df_scores_3. rename(columns={0:'Score'})
df_scores_3.sort_values(by='Score', ascending=False)

Score
random_forest  84.92
gaussian       82.68
logreg         82.12
decision_tree  79.89
knn            75.98
LinearSVC      65.92
SVC            65.36
perceptron     37.99
sgdclassifier  37.99

The score was not improved, the highest without changes was 86.76, the highest score with normalization was 85.47. So this has also not improved. The score, however, is better then the score with pca which was: 84.36.

4. Realiza 1, 2 y 3 pero usando GridSearch para encontrar los mejores hiperparámetros (features) de los algoritmos de ML antes de sacar el score. Debes probar con al menos 3 features y 2 valores para cada feature como mínimo.

- using gridsearch function imported from Utils.gridseach_model.py and if model was not there, made it and added it to function. 
- --> GaussianNB() does not have parameter tuning.

In [ ]:
4.1. Realizar una reducción de dimensionalidad/PCA para intentar aumentar el porcentaje de acierto. ¿Aumenta? ¿Cómo cambia el acierto según diferentes valores de PCA? Prueba, como mínimo, con valores para "n_components" 2,4,5 y 0.45 y 0.95. ¿Ha habido algún problema con algún valor?

Tried the following parameters for each n_component

n_2:
LogisticRegression(penalty='l1', solver='liblinear')
SVC(C=1, kernel='linear', probability=True)
KNeighborsClassifier(n_neighbors=9, weights='distance')
gaurssian -> no parameters
Perceptron(fit_intercept=False, shuffle=False)
LinearSVC(C=1, loss='hinge')
SGDClassifier(penalty='l1')
DecisionTreeClassifier(criterion='entropy')
RandomForestClassifier(criterion='entropy', n_estimators=50)

n_4:
LogisticRegression(penalty='none', solver='newton-cg')
SVC(C=1, kernel='linear', probability=True)
KNeighborsClassifier(n_neighbors=7, weights='distance')
gaurssian -> no parameters
Perceptron(penalty='l1', shuffle=False)
LinearSVC(C=8, loss='hinge')
SGDClassifier(early_stopping=True, loss='log', penalty='elasticnet')
DecisionTreeClassifier(min_samples_split=10, splitter='random')
RandomForestClassifier(n_estimators=200, warm_start=True)

n_5:
LogisticRegression(penalty='l1', solver='liblinear')
SVC(C=1, kernel='linear', probability=True)
KNeighborsClassifier(n_neighbors=7, weights='distance')
gaurssian -> no parameters
Perceptron(penalty='l2')
LinearSVC(C=1, loss='hinge')
SGDClassifier(loss='modified_huber', penalty='elasticnet')
DecisionTreeClassifier(min_samples_split=5, splitter='random')
RandomForestClassifier(criterion='entropy', max_features='sqrt',
                       n_estimators=200)

n_0.45:
--> gave lots of error and froze system. So did not do the 0.45.


n_0.95:
LogisticRegression(penalty='l1', solver='liblinear')
SVC(C=1, kernel='linear', probability=True)
KNeighborsClassifier(n_neighbors=9, weights='distance')
gaurssian -> no parameters
Perceptron(fit_intercept=False, shuffle=False)
LinearSVC(C=1, loss='hinge')
SGDClassifier()
DecisionTreeClassifier(criterion='entropy', min_samples_split=5,
                       splitter='random')
RandomForestClassifier(max_features='log2', n_estimators=200)





In [8]:

models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]

data = {"pca_2":[73.74, 75.42, 82.68, 36.31, 46.37, 75.42, 70.95, 82.12, 84.36], "pca_4":[81.01, 83.24, 82.68, 67.04, 72.63, 83.24, 46.93, 83.24, 83.8], "pca_5": [80.45, 82.12, 82.68, 69.27, 71.51, 82.12, 79.33, 82.68, 84.92], "pca_0.95": [73.74, 75.42, 82.68, 36.31, 46.37, 75.42, 65.36, 81.56, 83.8]}

df_scores_4 = pd.DataFrame(data, index = models)
df_scores_4.index.name = 'model'

df_scores_4.sort_values(by='pca_4', ascending=False)

pca_2  pca_4  pca_5  pca_0.95
model                                       
random_forest  84.36  83.80  84.92     83.80
SVC            75.42  83.24  82.12     75.42
LinearSVC      75.42  83.24  82.12     75.42
decision_tree  82.12  83.24  82.68     81.56
knn            82.68  82.68  82.68     82.68
logreg         73.74  81.01  80.45     73.74
perceptron     46.37  72.63  71.51     46.37
gaussian       36.31  67.04  69.27     36.31
sgdclassifier  70.95  46.93  79.33     65.36

Now highest score is with: 84.92. Still lower than the original score, but slightly higher then the highest score using pca without the parameters.   

4.2 Realizar una normalización de los datos no categóricos y comprobar si aumenta los scores.

parameters used:

LogisticRegression(solver='newton-cg')

nan --> SVC kept on running without doing anything, froze system. 

KNeighborsClassifier(n_neighbors=8)

Perceptron(penalty='l1', shuffle=False)

LinearSVC(C=1)

SGDClassifier(loss='log', penalty='elasticnet')

DecisionTreeClassifier(criterion='entropy', min_samples_split=10,
                       splitter='random')
                       
RandomForestClassifier(warm_start=True)

In [4]:
models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]

scores = [82.12, None, 87.15, 82.68, 65.92, 82.68, 72.07, 82.12, 84.92]

df_scores_5 = pd.DataFrame(scores, index = models)
df_scores_5 = df_scores_5. rename(columns={0:'Score'})
df_scores_5.sort_values(by='Score', ascending=False)

Score
knn            87.15
random_forest  84.92
gaussian       82.68
LinearSVC      82.68
logreg         82.12
decision_tree  82.12
sgdclassifier  72.07
perceptron     65.92
SVC              NaN

Knn has now the a better score than the original model: 87.15

4.3 Añade, como columnas, la media, mediana, varianza y covarianza de todas las columnas numéricas no categóricas y comprueba si con esas nuevas columnas se aumenta el score de los algoritmos.

parameters used:

LogisticRegression(solver='newton-cg')

2-> overgeslagen!

KNeighborsClassifier(n_neighbors=11, weights='distance')

Perceptron(shuffle=False)

LinearSVC(C=1, loss='hinge')

SGDClassifier(penalty='l1')

DecisionTreeClassifier(min_samples_split=5, splitter='random')

RandomForestClassifier(n_estimators=50, warm_start=True)









In [6]:
models = ["logreg", "SVC", "knn", "gaussian", "perceptron", "LinearSVC", "sgdclassifier", "decision_tree", "random_forest"]

scores = [82.12, 81.56, 77.09, 82.68, 65.92, 65.92, 81.01, 79.89, 83.8]


df_scores_6 = pd.DataFrame(scores, index = models)
df_scores_6 = df_scores_6. rename(columns={0:'Score'})
df_scores_6.sort_values(by='Score', ascending=False)

Score
random_forest  83.80
gaussian       82.68
logreg         82.12
SVC            81.56
sgdclassifier  81.01
decision_tree  79.89
knn            77.09
perceptron     65.92
LinearSVC      65.92

Highest score is:	83.80 with random forest. This is not a new highest score. And also it is worse than when we did not use any parameters. 

---------------------

Conclusion: the highest score is when normalizing the fare column and then using: KNeighborsClassifier(n_neighbors=8)   